Refactor compression code to expand URL support

dhimmel opened the pull request:
Work in progress. Picks up where #13340 left off.
 
closes #12688 
closes #13340 
closes #14570

Commit:
Move compression code to io.common._get_handle

Part of #14576.
Closes #13340.

jorisvandenbossche commented on pandas/io/common.py:
```diff
@@ -276,7 +276,7 @@ def file_path_to_url(path):
 # ZipFile is not a context manager for <= 2.6
 # must be tuple index here since 2.6 doesn't use namedtuple for version_info
-if sys.version_info[1] <= 6:
+if compat.PY2 and sys.version_info[1] <= 6:
```
We don't support python 2.6 anymore, so I think this check can be removed

dhimmel replied:
:+1: removed in 28f19f193dd6d1a924ad0e14996e47223df87738.

jorisvandenbossche commented on pandas/io/parsers.py:
```diff
@@ -1791,20 +1791,10 @@ def __init__(self, f, **kwds):
+        f = _get_handle(f, 'r', encoding=self.encoding,
+                        compression=self.compression,
+                        memory_map=self.memory_map)
+        self.handles.append(f)
```
This line (adding each `f` to ``handles``) will give problems, as only the handles we opened ourselves should be added, not handles passed by the user (previously those were only added inside the if statements)

jorisvandenbossche replied:
You can see the effect in the failure of the test_file_handles test

dhimmel replied:
@jorisvandenbossche I attempted a fix in 2f43c39e16b844a0f5552c407225387135d38e18, but it's not very elegant -- logic is repeated in ` _get_handle` and `PythonParser.__init__`.

What do you think is the best solution? What situation do we not end up opening a handle ourselves? Is it a problem that in PY3 we pass `path_or_buf` through `io.TextIOWrapper` before adding it to `handles`?

jorisvandenbossche reviewed (commented):
Can you add tests for the new features you added? (the compressed sources from urls)

jreback commented on pandas/io/common.py:
```diff
@@ -285,53 +285,84 @@ def ZipFile(*args, **kwargs):
 
-def _get_handle(path, mode, encoding=None, compression=None, memory_map=False):
-    """Gets file handle for given path and mode.
+def _get_handle(source, mode, encoding=None, compression=None,
```
call this ``path_or_buf`` in-line with spellings elsewhere

dhimmel replied:
Addressed in 673bde0503a854c9cce64ea98c4ac28d5071996f

dhimmel commented:
> Can you add tests for the new features you added? (the compressed sources from urls)

@jorisvandenbossche, do you want me to add xz/zip/bz2 compressed versions of `salary.table.csv` to [`pandas/io/tests/parser/data`](https://github.com/pandas-dev/pandas/tree/e1cdc4b83d7e970e9683056be722d5fa2a00fa70/pandas/io/tests/parser/data)? Also should I add my tests to [`pandas/io/tests/parser/test_network.py#L17-L35`](https://github.com/pandas-dev/pandas/blob/e1cdc4b83d7e970e9683056be722d5fa2a00fa70/pandas/io/tests/parser/test_network.py#L17-L35)?

dhimmel commented:
Are we also going to have to do something to [`common.py#L239-L244`](https://github.com/pandas-dev/pandas/blob/e1cdc4b83d7e970e9683056be722d5fa2a00fa70/pandas/io/common.py#L239-L244)?

jreback commented:
@dhimmel yes that should be more general

dhimmel commented on pandas/io/tests/parser/test_network.py:
```diff
@@ -27,51 +34,15 @@ def setUp(self):
-        url = self.base_url + '.zip'
-        url_table = read_table(url, compression="zip", engine="python")
-        tm.assert_frame_equal(url_table, self.local_table)
+    def test_compressed_urls(self):
```
@jorisvandenbossche, to try and reduce the repetition of the testing code, I made this commit (28373f1ecc0d3190dc18d98a86849ff9bbb0b841) which uses [nose test generators](http://nose.readthedocs.io/en/latest/writing_tests.html#test-generators). However, the nose doc states:

> Please note that method generators are not supported in unittest.TestCase subclasses.

As a result, the tests yielded by `test_compressed_urls` aren't actually running. I'm attempting to run the test using:

```sh
nosetests --verbosity=3 pandas/io/tests/parser/test_network.py:TestCompressedUrl.test_compressed_urls
```

Do you know the best way to proceed? I'm new to nose and pretty new to testing, so any advice will be appreciated.

jreback replied:
we run yielded test in computation and io/pickle

will be much easier when we switch to pytest

dhimmel replied:
Cool, I see the test generator examples in [`io/tests/test_pickle.py`](https://github.com/pandas-dev/pandas/blob/06b35db1d88e75a9e3f183b58678589cf7715381/pandas/io/tests/test_pickle.py#L194), [`computation/tests/test_compat.py`](https://github.com/pandas-dev/pandas/blob/06b35db1d88e75a9e3f183b58678589cf7715381/pandas/computation/tests/test_compat.py#L40), and [`computation/tests/test_eval.py`](https://github.com/pandas-dev/pandas/blob/06b35db1d88e75a9e3f183b58678589cf7715381/pandas/computation/tests/test_eval.py#L809). Those examples don't have a `setUp` method nor do they use the `pandas.util.testing.network` decorator. So I did my best in 0fcb3c328fcec8b1beb1ed8c968ed4c7835e5da1. It seems that one side effect is that the failure messages are less informative?

jreback commented on pandas/io/common.py:
```diff
@@ -274,64 +273,105 @@ def file_path_to_url(path):
+        from io import TextIOWrapper
+        return TextIOWrapper(path_or_buf, encoding=encoding)
+
+    elif compression:
```
can this be ``infer`` here? (IIRC we have an inference function based on the file ext). Is this guaranteed to be a valid compression? (its fine, though I think easy to handle it here anyhow).

dhimmel replied:
Currently, this function (`_get_handle`) cannot consume `compression=infer`. However, it would be an easy addition:

```python
if compression == 'infer':
    assert is_path
    compression = infer_compression(path_or_buff)
```

One issue I'm having is that `_get_handle` was undocumented before this PR. So it's a bit hard for me to know what to implement. @jreback or @jorisvandenbossche: can you comment on the tentative docstring and fill in the missing details, so I know exactly what to implement?

jorisvandenbossche replied:
@dhimmel What is still missing in your current docstring?

Adding `compression='infer'` here only makes sense if it would be used of course. Where happens the inference of the compression now? Because maybe it is already inferred before it is passed to this function?

dhimmel replied:
> Where happens the inference of the compression now?

@jorisvandenbossche, inferrence currently happens in [`pandas/io/parsers.py`](https://github.com/pandas-dev/pandas/blob/c045e1d6774aaa32ce13def79901f1a1ad8792bf/pandas/io/parsers.py#L357-L370).

I don't think `compression == 'infer'` will ever make it this far with the current design. @jreback does this answer your question?

jreback replied:
@dhimmel that's fine. I just want to validate somewhere that ``compression in [None, ......]`` or raise a ValueError

jreback commented on pandas/io/tests/parser/test_network.py:
```diff
@@ -14,24 +14,30 @@
-    def test_url_gz_infer(self):
-        url = 'https://s3.amazonaws.com/pandas-test/salary.table.gz'
-        url_table = read_table(url, compression="infer", engine="python")
+class TestCompressedUrl:
```
inherit from ``object``

jreback commented:
ideally if you can use some of this code in the parser routines would be great (may not be completely possible, but want to consolidate some of that duplication)

dhimmel commented:
Regarding [`common.py#L239-L244`](https://github.com/pandas-dev/pandas/blob/e1cdc4b83d7e970e9683056be722d5fa2a00fa70/pandas/io/common.py#L239-L244), I think there's a problem. Currently, when `compression="infer"` and the `path_or_buff` is a URL, `get_filepath_or_buffer` is using the [Content-Encoding](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Encoding) header of the request to infer the compression type. However, I don't actually think compressed files on the network will provide a Content-Encoding header of their compression type. Instead I think we want to infer the compression of a URL from it's name, just as we do with paths.

For example, `https://raw.githubusercontent.com/pandas-dev/pandas/master/pandas/io/tests/parser/data/salaries.csv.bz2` would infer `bz2` compression because it ends with `.bz2`. @jreback and @jorisvandenbossche, agree?

jreback commented:
i think inferring based on the file name is usually enough - but u can also use the Content if it's available

jreback commented on pandas/io/common.py:
```diff
@@ -274,64 +273,105 @@ def file_path_to_url(path):
-            f = lzma.LZMAFile(path, mode)
+            f = lzma.LZMAFile(path_or_buf, mode)
+
+        # Unrecognized Compression
```
ahh ok we are doing the validation, nvm. then. (assume we have some tests that hit this path)

dhimmel replied:
> assume we have some tests that hit this path

Actually, I don't think there were any tests for invalid compression (regardless of URL or not), so I added one. https://github.com/pandas-dev/pandas/pull/14576/commits/f2ce8f85b9a7d3d1000be9b3c09310aab67e38c6

jreback commented on pandas/io/parsers.py:
```diff
@@ -1791,19 +1791,17 @@ def __init__(self, f, **kwds):
-        # in Python 3, convert BytesIO or fileobjects passed with an encoding
-        elif compat.PY3 and isinstance(f, compat.BytesIO):
-            from io import TextIOWrapper
+        add_handle = (
```
*maybe* this could be a utility fnction in ``io/common`` (unless its not used anywhere else)

dhimmel replied:
I think `add_handle` is only used here -- I created it for this PR, so nowhere else depends on it.

I'm still a little unsure about the purpose of `self.handles` -- should it only contain file handles because right now we're adding handles that aren't files (for example in the case of `compat.PY3 and isinstance(f, compat.BytesIO)`)?

jreback replied:
the ``self.handles`` is to allow things to be closed if WE (meaning pandas opened) them, e.g. when you wrap with ``TextWrapper`` we NEED to close them. Otherwise we try to preserve open files handles (that were passed to us)

jorisvandenbossche replied:
Another way to do this would be to handle this inside `_get_handle` and return from that function `handles` that should be added. IMO that would be more maintainable, as now this boolean expression here depends on the content of the `_get_handle` function, so better to do it there.

jreback commented on pandas/io/tests/parser/test_network.py:
```diff
@@ -7,31 +7,44 @@
+        'gzip': '.gz',
+        'bz2': '.bz2',
+        'zip': '.zip',
+        'xz': '.xz',
```
is there a test for invalid compression specified?

dhimmel replied:
Did you want a specific test for invalid compression specified for a URL? As of the latest commit (bff0f3e2cdabfe02cfcbe95667a1a2ea1481fe1e), compression inference has been consolidated and invalid compression [raises an error](https://github.com/dhimmel/pandas/blob/bff0f3e2cdabfe02cfcbe95667a1a2ea1481fe1e/pandas/io/parsers.py#L374-L375) before compression is attempted.

dhimmel commented on pandas/io/common.py:
```diff
@@ -233,20 +199,16 @@ def get_filepath_or_buffer(filepath_or_buffer, encoding=None,
+        content_encoding = req.headers.get('Content-Encoding', None)
+        if content_encoding == 'gzip':
+            # Override compression based on Content-Encoding header
+            compression = 'gzip'
```
> i think inferring based on the file name is usually enough - but u can also use the Content if it's available

@jreback, `Content-Encoding` is not a versatile way to infer compression -- `gzip` is the only compression encoded that pandas and the Content-Encoding specification both support. I've modified it so inference on a URL uses its extension. However, if `Content-Encoding` is `gzip`, then compression is overridden. I don't expect this situation to arise very often. Does this make sense or should we just totally ignore `Content-Encoding`?

jreback replied:
yeah I agree with your assements. Since this *was* here, we can leave it. But I agree the filename inference is better / more standard.

dhimmel commented:
I'm getting [failing builds](https://ci.appveyor.com/project/jreback/pandas-465/build/1.0.2307/job/1f7g05x811genspf) in Python 2 due to `ImportError: No module named backports`. This occurs in `compat/__init__.py` at:

```python
    def import_lzma():
        """ import the backported lzma library
        or raise ImportError if not available """
        from backports import lzma
        return lzma
```

Any idea why `backports` is not available? How should I proceed?

jreback commented on pandas/io/common.py:
```diff
@@ -276,64 +237,105 @@ def file_path_to_url(path):
+            elif compat.PY2:
+                # Python 2's bz2 module can't take file objects, so have to
+                # run through decompress manually
+                f = StringIO(bz2.decompress(path_or_buf.read()))
```
I think you need to close the original file.

dhimmel replied:
By following with `path_or_buf.close()`?

jorisvandenbossche replied:
As this is a user provided file object, we shouldn't close it? @jreback or is this a special case because of the decompressing?

jreback commented on pandas/io/common.py:
```diff
@@ -276,64 +237,105 @@ def file_path_to_url(path):
+            f = open(path_or_buf, mode, encoding=encoding)
         else:
-            f = open(path, mode)
+            # Python 3 and no explicit encoding
```
I don't think you need all of this, just

``open(path_or_buf, mode, encoding=encoding)`` 

``encoding=None`` is valid on PY2.

we don't want to suppress issue with encoding, so we don't open with ``errors``

jreback replied:
oh, I c, we actually had this code from before...hmm. ok then.

jreback commented on pandas/io/common.py:
```diff
@@ -276,64 +237,105 @@ def file_path_to_url(path):
         else:
-            raise ValueError('Unrecognized compression type: %s' %
-                             compression)
+            msg = 'Unrecognized compression: {}'.format(compression)
```
I would preserve the same message (or even better expand it)

jreback commented on pandas/io/parsers.py:
```diff
@@ -348,38 +347,47 @@ def _validate_nrows(nrows):
+    for compression, extension in _compression_to_extension.items():
+        if filepath.endswith(extension):
+            return compression
+    return None
```
I would make this raise a nice ValueError if you have an *invalid* compress (accepting ``None/infer`` as valid). , then we can just call this.

jreback commented on pandas/io/parsers.py:
```diff
@@ -348,38 +347,47 @@ def _validate_nrows(nrows):
-            inferred_compression = None
+    compression = kwds.get('compression')
+    if compression not in set(_compression_to_extension) | {None, 'infer'}:
+        msg = 'Unrecognized compression type: {}'.format(compression)
```
you can do all of this in ``_infer_compression`` (which will raise if its a bad compression method), otherwise return the compression if ``infer`` (and just return if its None)

jorisvandenbossche replied:
To improve the error message, it would also be nice to list the possible values.

jreback commented on pandas/io/s3.py:
```diff
@@ -99,9 +99,7 @@ def get_filepath_or_buffer(filepath_or_buffer, encoding=None,
-    if compat.PY2 and (compression == 'gzip' or
-                       (compression == 'infer' and
-                        filepath_or_buffer.endswith(".gz"))):
+    if compat.PY2 and compression:
```
so you can just call ``_infer_compression`` I think in most of these cases

jorisvandenbossche commented on pandas/io/tests/parser/test_network.py:
```diff
@@ -7,31 +7,48 @@
+        for compression, extension in self.compression_to_extension.items():
+            url = self.base_url + extension
+            # args is a (compression, engine) tuple
+            for args in [(compression, 'python'), ('infer', 'python')]:
```
Is there a reason this is only tested for the python engine?

dhimmel replied:
I didn't think any of the changes I was making affected the `c` engine, since I didn't change any C code. But does the C engine kick in only after the file handle is opened? If so that's great and I'll add the C engine to the test. I thought someone was going to have to re-implement this feature changes in C.

jorisvandenbossche replied:
Not sure about that. Let's see what the test runs says, but if there are problems with the C engine, let's leave that for another PR. 
In any case, if those pass for the C engine, you can also use `self.read_csv` instead of `read_csv` and then the tests will automatically be run for both the python and C engine without needing to specify it explicitly

dhimmel commented:
@jorisvandenbossche, I added the `c` engine tests in https://github.com/pandas-dev/pandas/pull/14576/commits/e78d022949cd610276d83a820c45768840c25d0a. But we're gettings errors in the following [tests](https://travis-ci.org/pandas-dev/pandas/jobs/181193644):

```
ERROR: Test reading bz2-compressed tables from URL: compression="bz2", engine="c"
ERROR: Test reading bz2-compressed tables from URL: compression="infer", engine="c"
```

due to `ValueError: Python 2 cannot read bz2 from open file handle`. This error is coming from [`parser.pyx#L615-L621`](https://github.com/pandas-dev/pandas/blob/b3dd9bafbfcbac1a2b6ab756e588d6e167f90439/pandas/parser.pyx#L615-L621). This code looks to duplicate functionality of `_get_handle`, but is maybe Cython code? Anyways, I've never worked with `.pyx` code. Suggestions? Should we save `c` engine for a future PR?

jreback commented on pandas/io/parsers.py:
```diff
@@ -348,38 +347,47 @@ def _validate_nrows(nrows):
     return nrows
 
 
+_compression_to_extension = {
```
you can't sort None with strings in python 3.

```
 In [3]: set(['gz', 'xz']) | { None, 'infer'}
Out[3]: {None, 'xz', 'gz', 'infer'}

In [4]: sorted(set(['gz', 'xz']) | { None, 'infer'})
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-4-2dd21294aad8> in <module>()
----> 1 sorted(set(['gz', 'xz']) | { None, 'infer'})

TypeError: unorderable types: str() < NoneType()
```

instead you need to compare None separately (you can actually stringify things and it will work, e.g. None -> 'None') but that is non-pythonic

dhimmel replied:
Ahh. Okay fixed that but they're are these `pandas.io.common.ParserError`s that are killing me. https://github.com/dhimmel/pandas/commit/005b673eb924c30b84484331e36d23250b76dad9#commitcomment-20080291

jreback replied:
you prob need to rebuild the extensions. That is a relatively recent change.

dhimmel replied:
> you prob need to rebuild the extensions.

`python setup.py build_ext --inplace` did the trick!

https://github.com/pandas-dev/pandas/pull/14576/commits/a821b8207ea3a8ed54b3a79a7ae208da2adefc44 moves `_infer_compression` to `io/common.py`.

jreback commented on pandas/io/parsers.py:
```diff
@@ -348,38 +347,47 @@ def _validate_nrows(nrows):
+    is None. Otherwise, inference is perfomed using the extension of the
+    filename or URL.
+    """
+    if not isinstance(filepath_or_buffer, compat.string_types):
```
an FYI this would exclude pathlib type objects (but can deal with that later)

jorisvandenbossche commented:
Yes, let's leave issues with the C engine for another PR (also the pathlib stuff, there is an issue for it), and lets try to finish this PR.

dhimmel commented:
Still haven't migrated the handle tracking logic to `_infer_compression` as @jorisvandenbossche. The main thing holding me up is that I don't know exactly what should be considered a handle. Is this a prerequisite to merging this PR, or should we address later? 

> Yes, let's leave issues with the C engine for another PR

Okay c engine tests removed in https://github.com/pandas-dev/pandas/pull/14576/commits/bc9304389996496d92a72789c86b95b328e7925b. Prior to removing, the c-engine [tests](https://travis-ci.org/pandas-dev/pandas/builds/181502054) were all passing in 3. In 2.7 with the c-engine, the `bz2` reading from URL fails, but the other compression methods work.

I'd consider future work to add more S3 tests for all compression types. Also consolidating the [c-engine code](https://github.com/pandas-dev/pandas/blob/b3dd9bafbfcbac1a2b6ab756e588d6e167f90439/pandas/parser.pyx#L615-L621) that's making bz2 fail in 2.

When this PR is merged, it's important to me to have my authorship preserved in the commit history. I bring this up, since the first commit in this PR is from @lababidi -- I'm assuming they also care that their authorship is preserved. So just let me know if you'd like me to squash this thing in a certain way / how we should proceed.

Thanks for seeing this through. Can't wait to use the feature going forward and happy to keep working on these issues as needed.

jorisvandenbossche commented:
> The main thing holding me up is that I don't know exactly what should be considered a handle.

The handles we need to track, are all the handles (open files or file like objects) that pandas opens *itself* (not file handles provided by the user). You can check from the current `add_handle` boolean expression which ones it should be (those checks correspond to parts in the code of `_get_handle`. Tracking those opened handles inside `_get_handle` will make it a lot clearer what is going on.

I would like to see this fixed before merging

> I'd consider future work to add more S3 tests for all compression types. Also consolidating the c-engine code that's making bz2 fail in 2.

Can you open issues for those to keep track of them.

> When this PR is merged, it's important to me to have my authorship preserved in the commit history.

That's not problem, for sure! We can just squash your additional commits into one commit, and then commit those two commits of two authors to master.

dhimmel commented:
Will get around to finishing this early next week. Currently traveling.

jreback commented:
can you rebase, problem with appeveyor which i just fixed

dhimmel commented on pandas/io/common.py:
```diff
@@ -276,64 +237,144 @@ def file_path_to_url(path):
+        # In Python 3
         if compat.PY3:
             from io import TextIOWrapper
             f = TextIOWrapper(f, encoding=encoding)
```
So it looks like TextIOWrappers shouldn't count as distinct file objects. The following code closes `read_file` by `text.close()`.

```python
read_file = open(path, 'rb')
text = io.TextIOWrapper(read_file)
text.close()
```

dhimmel replied:
I addressed this by the following approach. If the object supports `.close()`, I considered it file-like and added it to `handles`. By returning a list (`handles`) containing new handles opened by `_get_handle`, we can return multiple new file objects. While you only need to close a single of these file objects to close them all, I assumed there's no harm in adding more handles than less.

jorisvandenbossche replied:
@dhimmel This looks like the appropriate approach!

dhimmel commented:
Assuming the tests pass for https://github.com/pandas-dev/pandas/pull/14576/commits/640dc1b0ed88bfdb9164083077d17af6a87ba826, I don't think there are any more unaddressed comments.

> can you rebase, problem with appeveyor which i just fixed

@jreback rebased with appveyor now passing.

> Can you open issues for those to keep track of them.

@jorisvandenbossche: I created #14874 and #14875.

> We can just squash your additional commits into one commit, and then commit those two commits of two authors to master.

@jorisvandenbossche do you want me to squash all of my commits? If so, what should go in the commit message? Should I add the [issues we're closing](https://github.com/pandas-dev/pandas/pull/14576#issue-187096549) to the commit message?

jreback commented:
@dhimmel you can squash or it will get squashed on the merge. in this case if you'd leave the original commit, and squash all of our subsequent ones to a single (or possibly) multiple, logically oriented commits would be great. We could in this case cherry-pick them (rather than completely squashing), to preserver authoriship (as I think when we squash ALL then authoriship is the last one). if we only have 2 commits a merge is easy.

dhimmel commented:
> in this case if you'd leave the original commit, and squash all of our subsequent ones to a single (or possibly) multiple, logically oriented commits would be great. 

I'm going to squash all of my commits. Grouping them will be pretty burdensome and the intermediate commits were often not functional. I [created a branch](https://github.com/dhimmel/pandas/commits/compression-unsquashed) on `dhimmel/pandas` to preserve the individual commits for reference. 

Just to list things here, `git log master.. --oneline` gives the following:

640dc1b Simplify TextIOWrapper usage
4774b75 Use list to store new handles
94aa0cb Track new handles in _get_handle https://github.com/pandas-dev/pandas/pull/14576#issuecomment-265283737 https://github.com/pandas-dev/pandas/pull/14576#discussion_r91988117
376dce3 Close bz2 file object in Python 2
b88d1b8 Remove compressed URL c-engine tests
b34c2f6 Move _infer_compression to io/common.py
d27e57d Fix "ImportError: No module named backports"
97a1c25 Test compressed URLs using c engine
ecee74f Revert to old error msg. Remove unused line
7c9fe8b Skip xz test if no lzma
492e615 Remove unused imports
bf9fe97 Add invalid compression test
7940ab7 Fix PEP8 errors
05332fd Update S3 get_filepath_or_buffer compression
24ea605 Fix get_filepath_or_buffer and _get_handle
2dc1f2f Clean up docstring
10652a0 Move compression inference to io/parsers
8d24bcf Delete unused function _wrap_compressed
83b2bc5 Infer compression from URL extension
c2e6e5b Improve error message, fix partial scope
9977da2 Fix nose test generator bug. Better error messages
7ff1247 Inheret from object for TestCompressedUrl
30e1caf Fix local salaries.csv path
6230dfb Update salaries.csv URL post #14587
6530f46 Enable nose execution of test generators
a441448 Broken: use method generators to DRY tests
25114c3 Tests for bz2, xz & zip URLs
89e45f4 Revert changes to zip error messages
29b7f74 Change error message in test_zip
3aba6cb Replicate original handles appending
101a344 Impove _get_handle docstring
58ffbce Rename source to path_or_buf and flake8
f0901fb Remove ZipFile support for Python 2.6
2c78178 Improve readability
592cf92 Fix ZipFile bug in Python 3.6+
74caf72 combine compression code into one

36 commits total in this PR. Wow!

dhimmel committed:
Refactor compression code to expand URL support

Part of #14576.
Closes #12688.
Closes #14570.
Opens #14874.

dhimmel commented:
> if we only have 2 commits a merge is easy.

Great we're now at 2 commits. I think merge makes sense, because it is nice to see the diff of both commits combined. Alone the changesets will be incomplete.

Let me know if there's anything more for me to do.

jorisvandenbossche commented:
@jreback You can use the github interface as well to "rebase and merge" instead of squashing, which will just apply the commits in this PR to master. And so the two commits now in the PR are perfect!

jreback commented:
@jorisvandenbossche right, but this is the exception to the rule.  I will have a look in a bit. 

@dhimmel going to merge, and can followup on comments / issues in another PR.

jreback commented:
thanks @dhimmel very nice!

in your xref PR, pls add whatsnew that cover all of the issues that were closed (for 0.20.0)

jorisvandenbossche commented:
@dhimmel thanks a lot!
